### ISSUE TITLE
Parallel work items should fail if the build has already failed

### ISSUE THREAD

**Issue by stmlange** [gradle/gradle#4296]
Hello,
I recently changed some of my code to use the new worker API (https://docs.gradle.org/4.1/userguide/custom_tasks.html#worker_api) and I noticed that the work done inside each worker might not be fully independent. Let me try to elaborate that on a scenario with three tasks:
- task1 and task2 can run in parallel and can be submitted to the worker api
- task3 depends on task1 and task2 and consumes generated files from both tasks, but only can run after those tasks have finished

Lets now assume I want to run task3 that will also invoke task1 and task2. Due to some missing files task1 fails its execution after 10 seconds (throwing a WorkerExecutionException), however task2 can still do its calculation and might be a long running task that runs for 30 minutes. I currently can't see a configuration / option that also lets task2 fail when any other submitted task fails. The main problem is that basically task2 can be aborted since we already know that our build fails and task3 should never get executed (since it needs some generated resources from task1 that are not available).
Is there such an option available?

### Expected Behavior

gradle as an option to propagate the failure to other tasks submitted to the worker-api and fails immediately when one of the worker failed.
However there are also cases where this behaviour might not be desired and thus gradle would need to ignore task failues in prallal work items for example when using ``--continue``. I think it would also be preferable to also have a programmatic way in telling gradle to ignore failures (e.g. ``gradle.workerApi.ignoreFailuresInWorkers = true`` inside ``build.gradle``).

### Current Behavior

gradle fails only when **all worker** submitted to the worker api has finished

### Sample
Just for reference I created a small example that is a slightly modified version of https://github.com/w25r/worker-api-example
In a nutshell it introduces a Sleep-task type that takes as arguments an amount of time it should sleep and if it should fail. The configuration of the tasks is the same as outlined above, but task 2 "just" sleeps for 30 seconds and task 1 fails after 10 seconds. This still shows the purpose of this feature request. The "work" done here is truly just sleeping, but that's not the point.

[issue-4296.zip](https://github.com/gradle/gradle/files/1707500/issue-4296.zip)

**Comment by blindpirate** [comment]
Could you please provide a reproducible sample? That would be great.

**Comment by stmlange** [comment]
Agree, if you use {{--continue}} this should still attempt to complete the other workers.
Just for reference I created a small example that is a slightly modified version of https://github.com/w25r/worker-api-example
In a nutshell it introduces a Sleep-task type that takes as arguments an amount of time it should sleep and if it should fail. The configuration of the tasks is the same as outlined above, but task 2 "just" sleeps for 30 seconds and task 1 fails after 10 seconds. This still shows the purpose of this feature request. The "work" done here is truly just sleeping, but that's not the point.

[issue-4296.zip](https://github.com/gradle/gradle/files/1707500/issue-4296.zip)

**Comment by MERGHOOB** [comment]
I am working on a big project which has 189 modules when I run the build from the top, if there is some compilation failure happens on some module, the build process does not stop.
I think the expected behavior should be like the --fast-fail feature for the test tasks. build should fail immediately after compilation failure or any other task failure encountered

**Comment by rgeorge-at-mentor** [comment]
We recently moved to gradle with a large number of modules and would like to see a compilation or any other task failure in any of the modules should fail fast, marking the build as failed.
